lib/logstash/filters/elasticsearch/client.rb

```diff
@@ -10,16 +10,14 @@ class ElasticsearchClient
 
       attr_reader :client
 
-      def initialize(user, password, options={})
+      def initialize(user, password, api_key, options = {})
```

Comment:
This method signature is getting increasingly complex and no longer follows actual usage, but this change is an opportunity to bring it back into alignment.

AFAICT, _at least one_ of the _required_ positional parameters needs to be `nil` (because supplying multiple forms of auth is an antipattern), and if one fails to provide the _optional_ `options` with just the right value, we'll end up crashing with a `NoMethodError` trying to `Enumerable#map!` over the missing `options[:hosts]` or trying to `Logger#info` the missing `options[:logger]`.

Since `hosts` and `logger` truly are the only required params, maybe we should re-arrange the signature to reflect that.

~~~
      def initialize(logger, hosts, options={})
        @logger = logger
        ssl = options.fetch(:ssl, false)
        user = options.fetch(:user, nil)
        password = options.fetch(:password, nil)
        api_key = options.fetch(:api_key, nil)
        
~~~

Reply:
make sense, agree.